ENH: LDLt decomposition for indefinite symmetric/hermitian matrices

ilayn opened the pull request:
Closes #6831 
Closes #6202


Since it seems that we are quite a few years away from the possibility of moving to LAPACK > 3.5.0. I've stripped down the rook and the aasen versions of ?he/sytrf routines. This at least provides one possibility which was missing as discussed in the linked issue, #7169 and https://github.com/numpy/numpy/pull/4078, https://github.com/numpy/numpy/pull/4079, https://github.com/numpy/numpy/pull/3960.

pv commented on scipy/linalg/tests/test_decomp_ldl.py:
```diff
@@ -0,0 +1,79 @@
+        lu_ind = tril_indices_from(x, k=-1)
+        # Test whether permutations lead to a triangular array
+        l, d, p = ldl(x, lower=0)
+        assert_(not any(l[p, :][lu_ind]), 'Spin {} failed'.format(_))
```
Also test `lower=True`?

ilayn replied:
Done

pv commented on scipy/linalg/tests/test_decomp_ldl.py:
```diff
@@ -0,0 +1,79 @@
+        assert_array_almost_equal(l.dot(d).dot(l.T)-x, zeros_like(x))
+
+    l, d, p = ldl(a, hermitian=False)
+    assert_array_almost_equal(l.dot(d).dot(l.T)-a, zeros_like(a))
```
Also test lower=True, hermitian=False

ilayn replied:
Done

pv commented on scipy/linalg/tests/test_decomp_ldl.py:
```diff
@@ -0,0 +1,79 @@
+               [3.91-1.50j, -1.78-1.18j, 0.11-0.11j, -1.84+0.00j]])
+    for x in (b, c, d):
+        l, d, p = ldl(x)
+        assert_array_almost_equal(l.dot(d).dot(l.T)-x, zeros_like(x))
```
Maybe simpler: `assert_allclose(l.dot(d).dot(l.T), x)`?

ilayn replied:
Done.

ilayn commented on scipy/linalg/tests/test_decomp_ldl.py:
```diff
@@ -0,0 +1,79 @@
+    # Complex matrix with imaginary diagonal entries with "hermitian=True"
+    with catch_warnings():
+        simplefilter('error')
+        assert_raises(ComplexWarning, ldl, A*1j)
```
Done

ilayn commented on scipy/linalg/_decomp_ldl.py:
```diff
@@ -0,0 +1,346 @@
+                    ldu[blk_i+x, blk_i+y], ldu[blk_i+x, blk_i+y].conj()]
+            else:
+                d[[blk_i+x, blk_i+y], [blk_i+y, blk_i+x]] = \
+                                                        ldu[blk_i+x, blk_i+y]
```
Done

ilayn commented on scipy/linalg/_decomp_ldl.py:
```diff
@@ -0,0 +1,346 @@
+            # should be conjugated.
+            if is_c and hermitian:
+                d[[blk_i+x, blk_i+y], [blk_i+y, blk_i+x]] = [
+                    ldu[blk_i+x, blk_i+y], ldu[blk_i+x, blk_i+y].conj()]
```
Done.

ilayn commented on scipy/linalg/_decomp_ldl.py:
```diff
@@ -0,0 +1,346 @@
+        # Not.
+        elif cur_val < 0 and cur_val == a[ind+x]:
+            # first neg entry of 2x2 block identifier
+            if -cur_val != ind+2:
```
It is taken care of by the line `x, y, rs, re, ri = (1, 0, 0, n, 1) if lower else (-1, -1, n-1, -1, -1)`. So for upper, the indices are going backwards and this is 2x2 branch so we jump over the element. It is indeed very confusing how LAPACK decided to encode this information. I hope I understand your remark correctly.

pv replied:
I'm asking because on the line above, the assignment is `swap_[ind+x] = swap_[-cur_val-1]` but the if statement checks for `-cur_val != ind+2` and not for `-cur_val != ind+x+1` so it's not analogous to the 1x1 case.

ilayn replied:
Yes, this is the 2x2 branch and for some reason this is how they decided to do it. The extra shift comes from the [LAPACK `ipiv`](http://www.netlib.org/lapack/explore-html/d3/db6/group__double_s_ycomputational_gad91bde1212277b3e909eb6af7f64858a.html#gad91bde1212277b3e909eb6af7f64858a) which really bothered me also for quite some time.

Quoting the relevant part 

>If `UPLO = 'U'` and `IPIV(k) = IPIV(k-1) < 0`, then rows and columns `k-1` and `-IPIV(k)` were interchanged and `D(k-1:k,k-1:k)` is a 2-by-2 diagonal block.  If `UPLO = 'L'` and `IPIV(k) =IPIV(k+1) < 0`, then rows and columns `k+1` and `-IPIV(k)` were interchanged and `D(k:k+1,k:k+1)` is a 2-by-2 diagonal block.

so not swap row k and IPIV(k) as was for 1x1 but the previous(next) one and IPIV(k).

ilayn commented on scipy/linalg/_decomp_ldl.py:
```diff
@@ -0,0 +1,346 @@
+        cur_val = a[ind]
+        # do we have a 1x1 block or not?
+        if cur_val > 0:
+            if cur_val != ind+1:
```
If I understand you correctly, `pivots`  in this branch is going to be set to 1 regardless because it is a 1x1 block. But if there is also a permutation it has to be recorded too. Otherwise we don't touch swap_ array.

pv replied:
I mean that
```
if cur_val != ind+1:
    swap_[ind] = swap_[cur_val-1]
```
is the same thing as
```
swap_[ind] = swap_[cur_val-1]
```
If the indices are the same, the assignment is a noop.

ilayn replied:
Ah I see what you mean. For that, I thought if the array is long enough then this would touch every element of the array with a read/write operation and waste time. So I only included the indices needed for swapping. If you think that's not a performance issue we can remove the `if` branch indeed.

ilayn commented on scipy/linalg/_decomp_ldl.py:
```diff
@@ -0,0 +1,346 @@
+    d : ndarray
+        The block diagonal multiplier of the factorization.
+    perm : ndarray
+        The permutation index array that brings lu into triangular form.
```
Done

ilayn commented on scipy/linalg/_decomp_ldl.py:
```diff
@@ -0,0 +1,346 @@
+        the factorization. Lower triangular (``lower=True``) is the default.
+    hermitian : bool, optional
+        In case the input matrix is complex valued, by setting this keyword
+        to False, the matrix is assumed to be only symmetric and not hermitian.
```
Reworded with a more explicit statement.

ilayn commented on scipy/linalg/_decomp_ldl.py:
```diff
@@ -0,0 +1,346 @@
+    matrix on entry would give the same result as if the full matrix is
+    supplied.
+
+    The permutation array can be used to triangulize the outer factors simply
```
Argh, yes.

ilayn commented on scipy/linalg/_decomp_ldl.py:
```diff
@@ -0,0 +1,346 @@
+    This function returns a block diagonal matrix D consisting blocks of size
+    at most 2x2 and also a possibly permuted unit lower triangular matrix
+    ``L`` such that the factorization ``A = L D L^H`` or ``A = L D L^T``
+    holds.
```
The default output matrices are possibly not triangular. And if I include those eqs. it gets a bit complicated. I've moved the permutation paragraph up and modified a bit more, to get the story hopefully more coherent. But let me know if you have a better wording please

ilayn commented on scipy/linalg/_decomp_ldl.py:
```diff
@@ -0,0 +1,346 @@
+from __future__ import division, absolute_import
```
Done.

ilayn commented:
Hope I've addressed the comments.

> I also wonder if writing completely separate branches for lower=True/False rather than using many small if statements everywhere would result to something clearer to read. If there's a bug in this code, modifying it later on will take effort.

The problem comes from the fact that the order is reversed in the lower/upper switch. The reason why I did as such is to keep things symmetric such that if there is a bug somewhere it automatically applies to both branches. The unpacking is really gross with this one.

ilayn commented:
The failing ones are from matplotlib and doctest of stats and signals.

pv commented on scipy/linalg/tests/test_decomp_ldl.py:
```diff
@@ -0,0 +1,90 @@
+    # Use upper part for the computation and use the lower part for comparison
+    l, d, p = ldl(e.conj().T, lower=0)
+    assert_allclose(tril(l.dot(d).dot(l.conj().T)-e), zeros((4, 4)),
+                    atol=spacing(10.))
```
The 10 ulp accuracy may be a bit tight --- putting maybe 1000 may save us having to revisit this later. Users may have linalg libraries with worse accuracy than what we have on the CI...

pv replied:
Ok, but indeed the default value for relative tolerance is `rtol==1e-7`.
Maybe this is not what you intended?

ilayn replied:
No problem, I'm being stupid here. I'll fix it.

pv commented on scipy/linalg/tests/test_decomp_ldl.py:
```diff
@@ -0,0 +1,90 @@
+
+
+def test_simple():
+    a = array([[-0.39-0.71j, 5.14-0.64j, -7.86-2.96j, 3.80+0.92j],
```
Reading this through again: add a comment how were these arrays generated.

pv replied:
Also, there does not appear to be clear reason to use predefined arrays, since they are not ground truth values. So maybe switch to random arrays?

Also, different data types are not tested. This would be simpler with random inputs.

ilayn replied:
I compiled it from other examples I've found around just to compare if it is working properly and did the random part with permutations but indeed I should add a proper random suite testing all dtypes and sizes. I'll add them.

ilayn commented:
Random matrix tests are added

pv commented:
great, thanks! merged.

ilayn commented:
Since now I can build SciPy on Windows, here is a simple speed benchmark for future reference. Blue is SciPy in the first plot

![image](https://user-images.githubusercontent.com/1303842/31640430-bdce4248-b2de-11e7-9942-b66ba37eba0a.png)


... asymptotically catching up matlab I guess

larsoner commented:
I'm pretty sure MATLAB uses MKL, you built against OpenBLAS right?

ilayn commented:
@larsoner Yes indeed.
